Improving "Connecting a Perpetual Stream with an HTTP Flow" section (fixes #504)

mzeltser69 opened the pull request:
fixes #504

akara commented on docs/streams-lifecycle.md:
````diff
@@ -153,6 +158,56 @@ class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUse
   override def streamGraph= MergeHub.source[MyMessage].to(Sink.ignore)
 }
 ```
+`PerpetualStreamWithMergeHub` input interface is `Sink[MyMessage, NotUsed]`, and no output.
````
I'm not sure using `input interface` in this context is right. `Sink[MyMessage, NotUsed]` is actually the materialized value of the `MergeHub`. So to be concise, I think of wording it along the following lines:

The materialized value of `MergeHub[MyMessage]` is `Sink[MyMessage, NotUsed]` allowing other streams to connect and dynamically materialize to this `MergeHub`. This means the `PerpetualStream` needs to expose this `Sink[MyMessage. NotUsed]` as one of the stream's materialized values for it to be useful to other streams.

akara commented on docs/streams-lifecycle.md:
```diff
@@ -153,6 +158,56 @@ class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUse
+When working with stream components it is important to keep in mind two separate prospective on the same code.
+From the outside our component have input and output interfaces. Input interface should be able to receive messages and hence needs to be a Sink.
+From the inside, the same input interface is the source of the messages, and it has type Sink.
+In this specific case our internal output points to nothing. In more complicated cases when we need to send messages out, our internal output will be `Sourse` from the outside prospective.
```
`Source`

akara commented on docs/streams-lifecycle.md:
```diff
@@ -153,6 +158,56 @@ class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUse
+
+Potentially confusing part when looking at this example is mixing `Sink` and `source` names to refer to the same. It looks a bit strange in English.
+When working with stream components it is important to keep in mind two separate prospective on the same code.
+From the outside our component have input and output interfaces. Input interface should be able to receive messages and hence needs to be a Sink.
```
I'd consider using different terms than input or output interfaces, except for very specific contexts. This term does not seem to be clearly defined in the context of Akka Streams.

mzeltser69 commented:
Adressed @akara comments on content and PR title.

anilgursel commented on docs/streams-lifecycle.md:
```diff
@@ -135,15 +135,18 @@ class MsgReceivingStream extends PerpetualStream[(ActorRef, Future[Done])] {
 
-Below is a sample `PerpetualStream` that uses `MergeHub`.
+Below are two sample `PerpetualStream` implementations, both use `MergeHub`.
+Type parameter `Sink[MyMessage, NotUsed]` describes the type of RunnableGraph that will be used as a destination (Sink) by the http flow part in `HttpFlowWithMergeHub` further down.
```
Can you please up RunnableGraph in back quotes like `RunnableGraph `.  Same for Sink.

anilgursel commented on docs/streams-lifecycle.md:
````diff
@@ -153,7 +156,57 @@ class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUse
 }
 ```
+From outside prospective (by http flow) this class is seen as terminal `Sink[MyMessage, NotUsed]`, which means that
+`PerpetualStreamWithMergeHub` expects to receive `MyMessage` and will not emmit anything out. 
````
typo in "emmit' ?

anilgursel commented on docs/streams-lifecycle.md:
```diff
@@ -153,7 +156,57 @@ class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUse
+Potentially confusing part when looking at this example is mixing `Sink` and `source` names to refer to the same. It looks a bit strange in English.
+Let's use outside vs. inside prospective explanation again.
+From the outside prospective our component seen as `Sink[MyMessage, NotUsed]`. This is achieved by using `MergeHub`, which from the inside prospective is a source of the messages hence val name `source`.
+Correspondingly, in the event we need to emmit something out, our val `sink` will be actually some shape of type `Source[MyMessage, NotUsed]`.
```
typo in "emmit".  Also, backquote on "val" please.

anilgursel commented on docs/streams-lifecycle.md:
```diff
@@ -173,7 +226,9 @@ squbs-actors = [
-The HTTP `FlowDefinition` can be connected to the `PerpetualStream` as follows by extending `PerpetualStreamMatValue` and using `matValue` method:
+The HTTP `FlowDefinition` can be connected to the `PerpetualStream` as follows by extending `PerpetualStreamMatValue` and using `matValue` method.
+Type parameter for the `PerpetualStreamMatValue` describes the materialized value of the expected receiving flow, which is our `PerpetualStreamWithMergeHub`.
+(both versions of `PerpetualStreamWithMergeHub` above expect to receive `MyMessage`, i.e. both are `Sink[MyMessage, NotUsed]`)
```
Shouldn't the dot be at the end, after the paranthesis?

anilgursel commented on docs/streams-lifecycle.md:
```diff
@@ -153,7 +156,57 @@ class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUse
+`MergeHub.source[MyMessage]` produces `Sink[MyMessage, NotUsed]` shape, which conforms to `PerpetualStream[Sink[MyMessage, NotUsed]]` type parameter.
+
+
+A bit more involved example using GraphDSL
```
Need a colon ":" at the end of the sentence:

anilgursel commented on docs/streams-lifecycle.md:
```diff
@@ -153,7 +156,57 @@ class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUse
+  val source = MergeHub.source[MyMessage]
+ 
+  //output interface - nothing
+  val sink = Sink.ignore
```
Since it a more involved example; however, doing a `Sink.foreach` and `println` ?  And making `flowGraph_enrichment ` actually appending a test to the message or modify it instead of doing `println`.

anilgursel commented on docs/streams-lifecycle.md:
```diff
@@ -153,7 +156,57 @@ class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUse
+  val sink = Sink.ignore
+  
+  //flow component which supposedly does something to MyMessage
+  val flowGraph_enrichment = Flow.fromFunction[MyMessage,MyMessage]{ x:MyMessage=>println(s"Event inside stream=$x");x}
```
`val` name should follow scala naming conventions.  I would write it as:

```scala
val flowGraph_enrichment = Flow.[MyMessage,MyMessage].map(message => Message(s"$message-enriched)}
```

anilgursel replied:
Yes, I was in a hurry 😄   Even though I mentioned `val` name, I did not put it my self 😄 

But, regarding the `Flow` itself, I think it would be a better example if we actually do an enrichment instead of a `println`.  I think `println` is better suited as `Sink.foreach(println)`.  This also makes the example better instead of doing a `Sink.ignore`.

akara commented on docs/streams-lifecycle.md:
````diff
@@ -153,7 +156,57 @@ class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUse
+```scala
+class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUsed]] {
+
+  override lazy val streamRunLifecycleState: LifecycleState = Initializing
````
There's a distinct reason for this line. Perhaps it needs some explanation (sorry Mike, that's a tough one).

akara commented on docs/streams-lifecycle.md:
```diff
@@ -153,7 +156,57 @@ class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUse
+  
+          import GraphDSL.Implicits._
+  
+          input ~> killSwitch.flow[MyMessage] ~> flowGraph_enrichment  ~> sink
```
One space too many before `  ~> sink`

akara commented on docs/streams-lifecycle.md:
````diff
@@ -153,7 +156,57 @@ class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUse
+```
+Let's see how all parts are falling into a place:
+`streamGraph` is expected to return `RunnableGraph` with the same type parameter as described in `PerpetualStream[Sink[MyMessage, NotUsed]]`.
+Our `source` is `MergeHub`, it is expected to receive `MyMessage`, which makes it `Sink[MyMessage,NotUsed]`
````
Our `source` is a `MergeHub`, it is expected to receive a `MyMessage`, which makes its materialized (runtime) type a `Sink[MyMessage, NotUsed]`**.**



*  A type relationship to an instance is generally referred to by the term "is a" as in "This message is a `MyMessage`.
*  Don't forget to add a period at the end of the sentence.

akara commented on docs/streams-lifecycle.md:
```diff
@@ -153,7 +156,57 @@ class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUse
+`streamGraph` is expected to return `RunnableGraph` with the same type parameter as described in `PerpetualStream[Sink[MyMessage, NotUsed]]`.
+Our `source` is `MergeHub`, it is expected to receive `MyMessage`, which makes it `Sink[MyMessage,NotUsed]`
+Our graph is built starting with our `source`, by passing it as a parameter to `GraphDSL.create(s:Shape)` constructor. 
+The result is `RunnableGraph[Sink[MyMessage, NotUsed]]` 
```
The result is a `RunnableGraph...`

or

The result is an instance of `RunnableGraph...`

akara commented on docs/streams-lifecycle.md:
```diff
@@ -153,7 +156,57 @@ class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUse
+The result is `RunnableGraph[Sink[MyMessage, NotUsed]]` 
+
+Potentially confusing part when looking at this example is mixing `Sink` and `source` names to refer to the same. It looks a bit strange in English.
+Let's use outside vs. inside prospective explanation again.
```
Perhaps a colon at the end of this sentence might make sense? What do think?

akara commented on docs/streams-lifecycle.md:
```diff
@@ -153,7 +156,57 @@ class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUse
+
+Potentially confusing part when looking at this example is mixing `Sink` and `source` names to refer to the same. It looks a bit strange in English.
+Let's use outside vs. inside prospective explanation again.
+From the outside prospective our component seen as `Sink[MyMessage, NotUsed]`. This is achieved by using `MergeHub`, which from the inside prospective is a source of the messages hence val name `source`.
```
out component **is** seen as...

mzeltser69 commented:
I have added better example - no fromFunction. Some dummy transformation, different input and output message types.
Fixed grammar and wording as requested - thank you for your patience!

akara commented on docs/streams-lifecycle.md:
```diff
@@ -153,7 +156,75 @@ class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUse
+`streamGraph` is expected to return `RunnableGraph` with the same type parameter as described in `PerpetualStream[Sink[MyMessage, NotUsed]]`.
+Our `source` is a `MergeHub`, it is expected to receive a `MyMessage`, which makes its materialized (runtime) type a `Sink[MyMessage,NotUsed]`.
+Our graph is built starting with our `source`, by passing it as a parameter to `GraphDSL.create(s:Shape)` constructor. 
+The result is an instance of `RunnableGraph[Sink[MyMessage, NotUsed]]`, which is a `ClosedShape` with `Sink` inlet and pluged outlet.
```
You probably meant "plugged outlet."

akara commented on docs/streams-lifecycle.md:
```diff
@@ -153,7 +156,75 @@ class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUse
+
+Potentially confusing part when looking at this example is mixing `Sink` and `source` names to refer to the same. It looks a bit strange in English.
+Let's use outside vs. inside prospective explanation again:
+From the outside prospective our component is seen as `Sink[MyMessage, NotUsed]`. This is achieved by using `MergeHub`, which from the inside prospective is a source of the messages hence `val` name `source`.
```
... is seen as a `Sink[MyMessage, NotUsed]`.

mzeltser69 commented:
removed fromGraph, used shorter names for flows

anilgursel commented on docs/streams-lifecycle.md:
```diff
@@ -153,7 +156,75 @@ class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUse
+  val sink = Sink.ignore
+  
+  //flow component, which supposedly does something to MyMessage
+  val preprocess = Flow[MyMessage].map{
```
I am not sure if this compiles. 

The signature is `def apply[T]: Flow[T, T, NotUsed]`; however, in this example, the `Flow` accepts a `MyMessage` and outputs a `MyMessageEnrich`.

mzeltser69 replied:
This is probably akka documentation error.
`val preprocess: Flow[MyMessage,MyMessageEnrich, NotUsed] =Flow[MyMessage].map{` 
 compiles and works.
I have  a real code in our internal github.

anilgursel commented on docs/streams-lifecycle.md:
```diff
@@ -153,7 +156,75 @@ class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUse
+  val enrichment = Flow[MyMessageEnrich].map{
+    inMsg=>
+      val outMsg = MyMessageEnrich(ip=inMsg.ip.replaceAll("\\.","-"), ts = inMsg.ts, enrichTs = System.currentTimeMillis()::inMsg.enrichTs)
+      println(s"Enriched Message inside enrich step=$outMsg")
```
I suggest removing `println` statement from the flow elements and instead do it in `Sink`, by `Sink.foreach(pritnln)`.

mzeltser69 replied:
Those println's are in each Flow to demo that logic in the Flow executed. If I move it out it is just printing end result. 
This not business logic just "debug" messages. 
Do you want me to add ActorLogging trait and wrap those messages in log.DEBUG, for clarity?

anilgursel commented on docs/streams-lifecycle.md:
```diff
@@ -153,7 +156,75 @@ class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUse
+
+  override lazy val streamRunLifecycleState: LifecycleState = Initializing
+  
+  // input interface - destination for MyMessage messages
```
I agree @akara's earlier comment.  "input interface" and "output interface" may cause confusions.

mzeltser69 replied:
fixing

anilgursel commented on docs/streams-lifecycle.md:
```diff
@@ -153,7 +156,75 @@ class PerpetualStreamWithMergeHub extends PerpetualStream[Sink[MyMessage, NotUse
+  
+    // building a flow based on another flow, to do some dummy enrichment
+  val enrichment = Flow[MyMessageEnrich].map{
+    inMsg=>
```
The convention in general is to keep this in the same line as `.map`.

mzeltser69 replied:
fixing

mzeltser69 commented:
all done
